funsor/terms.py

```diff
@@ -20,6 +20,63 @@
 from funsor.six import getargspec, singledispatch
 
 
+@dispatch(object, object)
+def substitute(expr, subs):
+    return expr  # default: do nothing
+
+
+@substitute.register(object, tuple)
+def substitute_with_tuple(expr, subs):
+    return substitute(expr, dict(subs))
```

Comment:
This introduces nondeterminism. I'd prefer to use tuples everywhere. What is your reason for using a dict?

Reply:
Nothing important, I'll change it to use tuples.